### PR TITLE
Inject build information during tool installation

### PR DESCRIPTION
**Pull request by cpanato**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

Inject build information during tool installation

the local build

```
$ docker run ba7873ce9e60 version
GitVersion:    v3.3.0-beta.0-1-g5a4c142-dirty
GitCommit:     5a4c142b979c7a4b9d359c1f05ed7250a83fc7a3
GitTreeState:  dirty
BuildDate:     '2021-10-04T14:21:25Z'
GoVersion:     go1.17.1
Compiler:      gc
Platform:      linux/amd64
```

/assign @justaugustus @puerco 
cc @kubernetes-sigs/release-engineering 

#### Which issue(s) this PR fixes:

Fixes: https://github.com/kubernetes-sigs/promo-tools/issues/436

#### Does this PR introduce a user-facing change?

```release-note
Inject build information during tool installation
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
+    BUILD_DATE ?= $(shell date "$(DATE_FMT)")
+endif
+GIT_TREESTATE = "clean"
+DIFF = $(shell git diff --quiet >/dev/null 2>&1; if [ $$? -eq 1 ]; then echo "1"; fi)
+ifeq ($(DIFF), 1)
+    GIT_TREESTATE = "dirty"
+endif
+
+PKG=sigs.k8s.io/promo-tools/v3/internal/version
+EXTRA_LDFLAGS='"-X $(PKG).gitVersion=$(GIT_VERSION) -X $(PKG).gitCommit=$(GIT_HASH) -X $(PKG).gitTreeState=$(GIT_TREESTATE) -X $(PKG).buildDate=$(BUILD_DATE)"'
```
Comment (justaugustus) [inline Makefile:38]:
```suggestion
LDFLAGS='"-X $(PKG).gitVersion=$(GIT_VERSION) -X $(PKG).gitCommit=$(GIT_HASH) -X $(PKG).gitTreeState=$(GIT_TREESTATE) -X $(PKG).buildDate=$(BUILD_DATE)"'
```

File: Makefile [hunk 2]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
+
+.PHONY: kpromo
+kpromo:
+	go build -trimpath -ldflags '-s -w -buildid= -extldflags "-static" $(EXTRA_LDFLAGS)' -o kpromo ./cmd/kpromo
```
Comment (justaugustus) [inline Makefile:42]:
```suggestion
	go build -trimpath -ldflags '-s -w -buildid= -extldflags "-static" $(LDFLAGS)' -o kpromo ./cmd/kpromo
```

File: Makefile [hunk 3]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
+
 ##@ Build
 .PHONY: build
-build: ## Go build
-	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/cmd/cip
+build: ## Go build cip-mm
```
Comment (justaugustus) [inline Makefile:46]:
```suggestion
build: ## Build go tools within the repository
```

File: Makefile [hunk 4]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
+	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/cmd/cip-mm
```
Comment (justaugustus) [inline Makefile:47]:
```suggestion
	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/cmd/cip-mm
	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/cmd/gh2gcs
	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/cmd/kpromo
```

File: Makefile [hunk 5]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
 	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/test-e2e/cip-auditor/cip-auditor-e2e.go
 	${REPO_ROOT}/go_with_version.sh build ${REPO_ROOT}/test-e2e/cip/e2e.go
 
 .PHONY: install
-install: build ## Install
-	cp cip $(shell go env GOPATH)/bin
+install: build ## Install cip-mm
```
Comment (justaugustus) [inline Makefile:52]:
```suggestion
install: build ## Install
```

File: Makefile [hunk 6]
```diff
@@ -19,16 +19,38 @@ REPO_ROOT:=$(shell dirname $(abspath $(lastword $(MAKEFILE_LIST))))
...
+	cp cip-mm $(shell go env GOPATH)/bin
```
Comment (justaugustus) [inline Makefile:53]:
```suggestion
	cp cip-mm $(shell go env GOPATH)/bin
	cp gh2gcs $(shell go env GOPATH)/bin
	cp kpromo $(shell go env GOPATH)/bin
```

We could opt to do a `go install` here instead.

File: Makefile (hunk 2 above)
Comment (justaugustus) [reply Makefile:42]: Should we add separate targets for `gh2gcs` and `cip-mm` as well?

**Review by justaugustus — CHANGES_REQUESTED** [review]
@cpanato -- Looks great so far!
While we're here, let's pull the `version` command up a level (since the version is true for all subcommands) i.e.,

```console
kpromo version
```

/retitle Inject build information during tool installation

**Comment by justaugustus** [comment]
Also:
/hold for https://github.com/kubernetes-sigs/promo-tools/pull/437

File: Makefile (hunk 2 above)
Comment (cpanato) [reply Makefile:42]: will do

Comment (cpanato) [reply Makefile]: thanks for the review and insights @justaugustus did all the changes PTAL

**Comment by cpanato** [comment]
```
$ docker run ba7873ce9e60 version
GitVersion:    v3.3.0-beta.0-1-g5a4c142-dirty
GitCommit:     5a4c142b979c7a4b9d359c1f05ed7250a83fc7a3
GitTreeState:  dirty
BuildDate:     '2021-10-04T14:21:25Z'
GoVersion:     go1.17.1
Compiler:      gc
Platform:      linux/amd64
```

**Review by justaugustus — APPROVED** [review]
Thanks @cpanato!
/hold cancel

**Pull request merged**